### DISCUSSION TITLE
Why is "msg_sender()" a Result?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [FuelLabs/sway#3430] [Q&A] [answered]
While browsing through the [examples](https://fuellabs.github.io/sway/v0.31.1/examples/subcurrency.html#subcurrency) in the docs, this piece of code caught my attention:

```sway
let sender: Result<Identity, AuthError> = msg_sender();
let sender = match sender.unwrap() {
    Identity::Address(addr) => addr,
    _ => revert(0),
};
```

Why is the `msg_sender` a Result and not always an `Identity::Address`? I presume this has to do with the FuelVM. If yes, in what cases is the `msg_sender` null?

For context, my background is in Solidity, and a little bit of Rust.

**Comment by otrho** [comment, marked as answer]
You can see here in the standard libarary where `msg_sender()` is implemented: https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/auth.sw#L34-L89

It can return `AuthError::InputsNotAllOwnedBySameAddress` if the caller is external.  I'm not an application dev so I can't elaborate more on the full semantics there, but it should make sense?

**Reply by PaulRBerg** [reply]
Thanks Toby, this is helpful.

I am still unsure in what cases one would bump into the `InputsNotAllOwnedBySameAddress` error.

Does that happen when the signature is forged and one of the input UTXOs do not belong to the same caller?

**Comment by mohammadfawaz** [comment]
> Does that happen when the signature is forged and one of the input UTXOs do not belong to the same caller?

Exactly. The inputs need to belong to the same caller which could be an "address" or another "contract", hence the type `Identity` wrapped in `Result`.
